File: 300/399 - Evaluate Division.java
import java.util.HashMap;

public class Solution {
    
    HashMap<String, HashMap<String, Double>> divs;
    HashSet<String> isVisited;
    
    public double[] calcEquation(String[][] equations, double[] values, String[][] queries) {
        int i, count = 0;
        String a, b;
        double d;
        double[] ret = new double[queries.length];
        
        this.divs = new HashMap<String, HashMap<String, Double>>();
        this.isVisited = new HashSet<String>();
        
        for (i = 0; i < equations.length; ++i) {
            a = equations[i][0];
            b = equations[i][1];
            d = values[i];
            
            if (! this.divs.containsKey(a)) {
                this.divs.put(a, new HashMap<String, Double>());
                this.divs.get(a).put(a, 1.0);
            }
            
            if (! this.divs.containsKey(b)) {
                this.divs.put(b, new HashMap<String, Double>());
                this.divs.get(b).put(b, 1.0);
            }
            
            this.divs.get(a).put(b, d);
            if (d != 0.0)
                this.divs.get(b).put(a, 1.0 / d);
        }
        
        for (String[] q : queries) {
            ret[count++] = dfs(q[0], q[1]);
            this.isVisited.clear();
        }
        
        return ret;
    }
    
    private double dfs(String a, String b) {
        double ret = -1.0, temp;
        HashMap<String, Double> hm;

        this.isVisited.add(a);
        if (! this.divs.containsKey(a)) {
            this.isVisited.remove(a);
            return ret;
        }
        hm = this.divs.get(a);
        
        for (String key : hm.keySet()) {
            if (b.equals(key)) {
                ret = hm.get(b);
                break;
            } else if (! this.isVisited.contains(key)) {
                if ((temp = dfs(key, b)) != -1.0) {
                    ret = hm.get(key) * temp;
                    break;
                }
            }
        }
            
        this.isVisited.remove(a);
        return ret;
    }
}

File: csharp/399.evaluate-division.cs
/*
 * @lc app=leetcode id=399 lang=csharp
 *
 * [399] Evaluate Division
 */

using System.Collections.Generic;

public class Solution {

    private double CalcEquationRecurse(IDictionary<string, IDictionary<string, double>> map, ISet<string> visited, string num, string denom) {
        if (map[num].ContainsKey(denom)) {
            return map[num][denom];
        }
        foreach (var entry in map[num]) {
            string next = entry.Key;
            if (visited.Contains(next)) {
                continue;
            }
            visited.Add(next);
            double val = CalcEquationRecurse(map, visited, next, denom);
            if (val > 0) {
                return val * entry.Value;
            }
        }
        return -1.0;
    }

    public double[] CalcEquation(IList<IList<string>> equations, double[] values, IList<IList<string>> queries) {
        IDictionary<string, IDictionary<string, double>> map = new Dictionary<string, IDictionary<string, double>>();
        double[] ret = new double[queries.Count];
        for (int i = 0; i < equations.Count; ++i) {
            var equation = equations[i];
            double val = values[i];
            string num = equation[0], denom = equation[1];
            if (!map.ContainsKey(num)) {
                map[num] = new Dictionary<string, double>();
            }
            if (!map.ContainsKey(denom)) {
                map[denom] = new Dictionary<string, double>();
            }
            map[num][denom] = val;
            map[denom][num] = 1.0 / val;
        }
        for (int i = 0; i < queries.Count; ++i) {
            var query = queries[i];
            string num = query[0], denom = query[1];
            if (!map.ContainsKey(num) || !map.ContainsKey(denom)) {
                ret[i] = -1.0;
            } else {
                ret[i] = CalcEquationRecurse(map, new HashSet<string>(), num, denom);
            }
        }
        return ret;
    }
}

